Depdetail

foursquare-markcc opened the pull request:
```
Add dependency checking to scala_compile.

Adds an optional analysis pass post-compilation for scala
targets that uses real class dependencies to generate a list
of the scala targets that any compilation target actually depends
on, and then compares that to the declared list of dependencies
to detect any spurious dependencies that are accidentally permitted
by the chunked compiler.
```

foursquare-markcc committed:
Add dependency checking to scala_compile.

Adds an optional analysis pass post-compilation for scala
targets that uses real class dependencies to generate a list
of the scala targets that any compilation target actually depends
on, and then compares that to the declared list of dependencies
to detect any spurious dependencies that are accidentally permitted
by the chunked compiler.

The new analysis is only run if the flag "check_missing_deps" is
passed to the scala compiler. (e.g., by running "pants compile --compile-scalac-check_missing_deps").

foursquare-markcc committed:
Add dependency checking to scala_compile.

Adds an optional analysis pass post-compilation for scala
targets that uses real class dependencies to generate a list
of the scala targets that any compilation target actually depends
on, and then compares that to the declared list of dependencies
to detect any spurious dependencies that are accidentally permitted
by the chunked compiler.

The new analysis is only run if the flag "check_missing_deps" is
passed to the scala compiler. (e.g., by running "pants compile --compile-scalac-check_missing_deps").

foursquare-markcc committed:
Extensions and cleanups for dep checking.

- Dep checking has been moved into a new class.
- Now, reasonable errors are generated saying why an undeclared dep exists,
  by showing a concrete examplar demonstrating a source file depending on a class
  from the undeclared target dependency.

foursquare-markcc committed:
Extensions and cleanups for dep checking.

- Dep checking has been moved into a new class.
- Now, reasonable errors are generated saying why an undeclared dep exists,
  by showing a concrete examplar demonstrating a source file depending on a class
  from the undeclared target dependency.

benjyw commented on src/python/twitter/pants/tasks/scala_compile.py:
```diff
@@ -71,6 +72,18 @@ def setup_parser(cls, option_group, args, mkflag):
+                            default=False,
+                            help="[%default] Check for undeclared dependencies in scala code")
+
+    option_group.add_option(mkflag("check_all_deps"), mkflag("check_all_deps", negate=True),
```
Not sure I understand the distinction between check_missing_deps and check_all_deps

foursquare-markcc replied:
I should probably come up with better names.

check_missing_deps is the toggle for controlling the entire deps analysis.
If you don't pass that, it completely skips the missing dependencies
analysis.

check_all_deps is a flag for the missing deps analysis. Without it, it will
omit warnings for classes that aren't declared in a scala target. So if you
have an undeclared dep on a java class, without check_all_deps, it won't
tell you.

At the moment, check_all_deps is only useful for debugging. When I do the
analysis, I don't look at jar dependencies - so if a class depends on
something that it's getting from a jar library, rather than  from scala
source, the analysis won't see it.

```
-mark
```

On Tue, Nov 6, 2012 at 6:38 PM, Benjy Weinberger
notifications@github.comwrote:

> In src/python/twitter/pants/tasks/scala_compile.py:
> 
> > @@ -71,6 +72,18 @@ def setup_parser(cls, option_group, args, mkflag):
> >                              action="callback", callback=mkflag.set_bool,
> >                              help="[True] Enable color in logging.")
> > -    option_group.add_option(mkflag("check_missing_deps"), mkflag("check_missing_deps", negate=True),
> > -                            dest="scala_check_missing_deps",
> > -                            action="callback", callback=mkflag.set_bool,
> > -                            default=False,
> > -                            help="[%default] Check for undeclared dependencies in scala code")
> >   +
> > -    option_group.add_option(mkflag("check_all_deps"), mkflag("check_all_deps", negate=True),
> 
> Not sure I understand the distinction between check_missing_deps and
> check_all_deps
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/foursquare/twitter-commons/pull/3/files#r2050146.

## 

Mark Craig Chu-Carroll
**\* Software Engineer@foursquare, Software Tools/Programming Language/Math
Geek
**\* Email: markcc@foursquare.com markcc@gmail.com; Twitter: @MarkCC
**\* Blog: http://scientopia.org/blogs/goodmath

benjyw replied:
Can we boil this down to one flag? This is really confusing. Surely we do want to check java/jar deps as well as scala deps?

foursquare-markcc replied:
It's a temporary placeholder. I don't want to remove or totally disable the
code; but I don't want to pollute the interface that people will really use.

Of course, eventually, we want to check everything - scala source deps,
java deps, jar deps, and anything else that could produce Java bytecodes.
But this is a first cut. As it stands, running only scala dep checkns, this
provides useful information, which is why I pushed it. Once the code to
extract the necessary information from java targets and jars is ready, then
I'm planning to remove the "check_all_deps" option entirely - if you want
to check if all deps are declared, you should be told about _all_
undeclared deps.

How about, as an interim measure, give the flag a name that makes it clear
that it's a temporary debugging aid? "debug_dep_checker"?

```
 -Mark
```

On Tue, Nov 6, 2012 at 6:59 PM, Benjy Weinberger
notifications@github.comwrote:

> In src/python/twitter/pants/tasks/scala_compile.py:
> 
> > @@ -71,6 +72,18 @@ def setup_parser(cls, option_group, args, mkflag):
> >                              action="callback", callback=mkflag.set_bool,
> >                              help="[True] Enable color in logging.")
> > -    option_group.add_option(mkflag("check_missing_deps"), mkflag("check_missing_deps", negate=True),
> > -                            dest="scala_check_missing_deps",
> > -                            action="callback", callback=mkflag.set_bool,
> > -                            default=False,
> > -                            help="[%default] Check for undeclared dependencies in scala code")
> >   +
> > -    option_group.add_option(mkflag("check_all_deps"), mkflag("check_all_deps", negate=True),
> 
> Can we boil this down to one flag? This is really confusing. Surely we do
> want to check java/jar deps as well as scala deps?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/foursquare/twitter-commons/pull/3/files#r2050348.

## 

Mark Craig Chu-Carroll
**\* Software Engineer@foursquare, Software Tools/Programming Language/Math
Geek
**\* Email: markcc@foursquare.com markcc@gmail.com; Twitter: @MarkCC
**\* Blog: http://scientopia.org/blogs/goodmath

benjyw commented on src/python/twitter/pants/tasks/scala_compile.py:
```diff
@@ -71,6 +72,18 @@ def setup_parser(cls, option_group, args, mkflag):
                             help="[True] Enable color in logging.")
 
+    option_group.add_option(mkflag("check_missing_deps"), mkflag("check_missing_deps", negate=True),
+                            dest="scala_check_missing_deps",
```
Flag names should use dashes:

mkflag("check-missing-deps") etc.

even though the dest variable naturally uses underscores.

Commit:
fixes no-color option on cmdline.

passing --no-compile-scala-color inverts the first flag, so the
or was resolving to `true` not `false` as needed.

benjyw committed:
Merge pull request #5 from jcrobak/fix-no-color

fixes no-color option on cmdline.

jcrobak committed:
update ant/templates/ivy.mk to fix test.

Had to update both ivy.mk files, since one is uses for
doc generation.

benjyw committed:
Merge pull request #6 from jcrobak/fix-doc-test

update ant/templates/ivy.mk to fix test.

foursquare-markcc committed:
Switch dep checking to use Twitter classfile parser.

foursquare-markcc committed:
Switch dep checking to use Twitter classfile parser.

foursquare-markcc committed:
Merge branch 'depdetail' of github.com:foursquare/twitter-commons into depdetail
